chia/data_layer/data_store.py

```diff
@@ -0,0 +1,675 @@
+        async with self.db_wrapper.locked_transaction(lock=lock):
+            generation = await self.get_tree_generation(tree_id=tree_id, lock=False)
+            cursor = await self.db.execute(
+                "SELECT * FROM root WHERE tree_id == :tree_id AND generation == :generation",
+                {"tree_id": tree_id.hex(), "generation": generation},
+            )
+            [root_dict] = [row async for row in cursor]
+
+        return Root.from_row(row=root_dict)
+
+    async def get_ancestors(self, node_hash: bytes32, tree_id: bytes32, *, lock: bool = True) -> List[InternalNode]:
+        async with self.db_wrapper.locked_transaction(lock=lock):
+            root = await self.get_tree_root(tree_id=tree_id, lock=False)
+            assert root.node_hash
+            assert root  # todo handle errors
+            cursor = await self.db.execute(
+                """
+                WITH RECURSIVE
+                    tree_from_root_hash(hash, node_type, left, right, key, value, depth) AS (
+                        SELECT node.*, 0 AS depth FROM node WHERE node.hash == :root_hash
+                        UNION ALL
+                        SELECT node.*, tree_from_root_hash.depth + 1 AS depth FROM node, tree_from_root_hash
+                        WHERE node.hash == tree_from_root_hash.left OR node.hash == tree_from_root_hash.right
+                    ),
+                    ancestors(hash, node_type, left, right, key, value, depth) AS (
+                        SELECT node.*, NULL AS depth FROM node
+                        WHERE node.left == :reference_hash OR node.right == :reference_hash
+                        UNION ALL
+                        SELECT node.*, NULL AS depth FROM node, ancestors
+                        WHERE node.left == ancestors.hash OR node.right == ancestors.hash
+                    )
+                SELECT * FROM tree_from_root_hash INNER JOIN ancestors
+                WHERE tree_from_root_hash.hash == ancestors.hash
+                ORDER BY tree_from_root_hash.depth DESC
+                """,
+                {"reference_hash": node_hash.hex(), "root_hash": root.node_hash.hex()},
+            )
+
+            # The resulting rows must represent internal nodes.  InternalNode.from_row()
+            # does some amount of validation in the sense that it will fail if left
+            # or right can't turn into a bytes32 as expected.  There is room for more
+            # validation here if desired.
+            ancestors = [InternalNode.from_row(row=row) async for row in cursor]
+
+        return ancestors
+
+    async def get_pairs(self, tree_id: bytes32, *, lock: bool = True) -> List[TerminalNode]:
+        async with self.db_wrapper.locked_transaction(lock=lock):
+            root = await self.get_tree_root(tree_id=tree_id, lock=False)
+
+            if root.node_hash is None:
+                return []
+
+            cursor = await self.db.execute(
+                """
```

Comment:
@altendky getting this " sqlite3.OperationalError: table tree_from_root_hash has 9 values for 8 columns"
after adding the status column for all these type of queries 
should we use specific column names instead of node.*,?

whats the correct way of handling this here ?

Reply:
see [`e4a67e9` (#8828)](https://github.com/Chia-Network/chia-blockchain/pull/8828/commits/e4a67e9be6f440e556deb3a5fc5726378789215d)